java/vector/src/main/codegen/templates/DenseUnionVector.java

```diff
@@ -0,0 +1,907 @@
+ */
+public class DenseUnionVector implements FieldVector {
+
+  private String name;
+  private BufferAllocator allocator;
+  int valueCount;
+
+  NonNullableStructVector internalStruct;
+  private ArrowBuf typeBuffer;
+  private ArrowBuf offsetBuffer;
+  private ArrowBuf validityBuffer;
+
+  /**
+   * The key is type Id, and the value is vector.
+   */
+  private ValueVector[] childVectors = new ValueVector[Byte.MAX_VALUE + 1];
+
+  /**
+   * The index is the type id, and the value is the type field.
+   */
+  private Field[] typeFields = new Field[Byte.MAX_VALUE + 1];
+
+  /**
+   * The next typd id to allocate.
+   */
+  private byte nextTypeId = 0;
+
+  private FieldReader reader;
+
+  private final CallBack callBack;
+  private long typeBufferAllocationSizeInBytes;
+  private long offsetBufferAllocationSizeInBytes;
+  private long validityBufferAllocationSizeInBytes;
+
+  private final FieldType fieldType;
+
+  private static final byte TYPE_WIDTH = 1;
+  public static final byte OFFSET_WIDTH = 4;
+
+  private static final FieldType INTERNAL_STRUCT_TYPE = new FieldType(false /*nullable*/,
+          ArrowType.Struct.INSTANCE, null /*dictionary*/, null /*metadata*/);
+
+  public static DenseUnionVector empty(String name, BufferAllocator allocator) {
+    FieldType fieldType = FieldType.nullable(new ArrowType.Union(
+            UnionMode.Dense, null));
+    return new DenseUnionVector(name, allocator, fieldType, null);
+  }
+
+  public DenseUnionVector(String name, BufferAllocator allocator, FieldType fieldType, CallBack callBack) {
+    this.name = name;
+    this.allocator = allocator;
+    this.fieldType = fieldType;
+    this.internalStruct = new NonNullableStructVector("internal", allocator, INTERNAL_STRUCT_TYPE,
+            callBack);
+    this.validityBuffer = allocator.getEmpty();
+    this.validityBufferAllocationSizeInBytes =
+        DataSizeRoundingUtil.divideBy8Ceil(BaseValueVector.INITIAL_VALUE_ALLOCATION);
+    this.typeBuffer = allocator.getEmpty();
+    this.callBack = callBack;
+    this.typeBufferAllocationSizeInBytes = BaseValueVector.INITIAL_VALUE_ALLOCATION * TYPE_WIDTH;
+    this.offsetBuffer = allocator.getEmpty();
+    this.offsetBufferAllocationSizeInBytes = BaseValueVector.INITIAL_VALUE_ALLOCATION * OFFSET_WIDTH;
+  }
+
+  public BufferAllocator getAllocator() {
+    return allocator;
+  }
+
+  @Override
+  public MinorType getMinorType() {
+    return MinorType.UNION;
+  }
+
+  @Override
+  public void initializeChildrenFromFields(List<Field> children) {
+    internalStruct.initializeChildrenFromFields(children);
+  }
+
+  @Override
+  public List<FieldVector> getChildrenFromFields() {
+    return internalStruct.getChildrenFromFields();
+  }
+
+  @Override
+  public void loadFieldBuffers(ArrowFieldNode fieldNode, List<ArrowBuf> ownBuffers) {
+    if (ownBuffers.size() != 3) {
+      throw new IllegalArgumentException("Illegal buffer count, expected " + 2 + ", got: " + ownBuffers.size());
+    }
+
```

Comment:
should this reference [typeIds metadata data someplace?](https://github.com/apache/arrow/blob/master/format/Schema.fbs#L97)

Reply:
Sure. Revised the exception message.